Updated docs (#29)
umbraco/Umbraco.AuthorizedServices: README.md

## Aims

**Umbraco Authorized Services** is an Umbraco package designed to reduce the effort needed to integrate third party services that require authentication and authorization via an OAuth or API key based flow into Umbraco solutions.  It's based on the premise that working with these services requires a fair bit of plumbing code to handle creating an authorized connection.  This is necessary before the developer working with the service can get to actually using the provided API to implement the business requirements.

Having worked with a few OAuth integrations across different providers, as would be expected, there are quite a few similarities to the flow that needs to be implemented.  Steps include:

- Redirecting to an authentication endpoint.
- Handling the response including an authentication code (or an oauth token and verifier code) and exchanging it for an access token.
- Securely storing the token.
- Including the token in API requests.
- Serializing requests and deserializing the API responses.
- Handling cases where the token has expired and obtaining a new one via a refresh token.

> For API key based flows, the integration steps include:
> 
> - Use an API key stored in the settings file or persisted into the database to serialize requests and deserialize the API responses.
> - Use provisioning options for passing the key to the API.

Reviewer: The intention of this section is more about the challenges that the package helps overcome, rather than detail how we've gone about solving it.  So I'd suggest:

```suggestion
With API key based flows, the process is a little simpler.  But you still have to consider secure storage of the key, providing it correctly in API calls and handling serialization tasks.
```

        {
          "<serviceAlias>": {
            "DisplayName": "",

>             "AuthenticationMethod": "",

Reviewer: We should set the different options of the enum here separated by vertical bars I think, just to be consistent with what we have for `true|false` below.  Think it's also useful to see at a glance what the options are.

```suggestion
            "AuthenticationMethod": "OAuth2AuthorizationCode|OAuth2ClientCredentials|OAuth1|ApiKey",
```

Provides a friendly name for the service used for identification in the user interface.

###### AuthenticationMethod

Specifies the type of authentication the service will use, from one of the following available options: `OAuth1`, `OAuth2AuthorizationCode`, `OAuth2ClientCredentials` or `ApiKey`.

> If none provided, it will default to `OAuth2AuthorizationCode`.

Reviewer: ```suggestion
If no value is provided, it will default to `OAuth2AuthorizationCode`.
```

###### ClientCredentialsProvision

> Specifies the available options for providing credentials in an `OAuth2` flow: `AuthHeader` or `RequestBody`.

Reviewer: Think we should add something here about what `AuthenticatedMethod` this is relevant for.
```suggestion
Specifies the available options for providing credentials in an `OAuth2` flow: `AuthHeader` or `RequestBody`.

This setting is only utilized when the `AuthenticatedMethod` value is configured as `OAuth2ClientCredentials`.
```

Used, along with `IdentityHost` to construct a URL that the user is redirected to when initiating the authorization of the service via the backoffice. For GitHub, the required value is `/login/oauth/authorize`.

###### CanManuallyProvideToken

> Specifies whether an administrator can manually add tokens.

Reviewer: ```suggestion
Specifies whether the service supports generating of tokens via the provider's developer portal such that an administrator can manually add one via the backoffice. 
```

###### CanManuallyProvideApiKey

> Specifies whether an administrator can manually add API keys.

Reviewer: ```suggestion
Specifies whether an administrator can manually add API keys via the backoffice. You might prefer to use this option instead of storing the key in configuration via the `ApiKey` setting.
```

###### CanExchangeToken

Specifies whether the access token can be exchanged with a long lived one.

###### ExchangeTokenProvision

> Provides a strongly typed configuration for a setup that allows exchanging an access token. 

Reviewer: ```suggestion
Provides a strongly typed configuration for a setup that allows exchanging an access token. 

This setting is only utilized when `CanExchangeToken` is set to `true`.

```

> In this case the configuration includes:

Reviewer: ```suggestion
The configuration of exchange tokens includes:

```

This flag indicates whether the basic token should be included in the request for access token. If true, a base64 encoding of <clientId>:<clientSecret> will be added to 
the authorization header.

###### API Key

> Specifies the key a service with `AuthenticationMethod=ApiKey` will use for making authorized requests to the API.

Reviewer: ```suggestion
Specifies the key a service with `AuthenticationMethod` set to `ApiKey` will use for making authorized requests to the API.
```
